### ISSUE TITLE
PNG files changing/becoming corrupt after I finished working on them

### ISSUE THREAD

**Issue by rodrigo1406** [rstudio/rstudio#11798]
### System details

    RStudio Edition : Desktop
    RStudio Version : 2022.07.1 Build 554
    OS Version      : Debian GNU/Linux 11.3 (bullseye) 64-bit
    R Version       : 4.0.4 (2021-02-15) -- "Lost Library Book"

### Steps to reproduce the problem

It happes by chance. Cannot reproduce it.

### Describe the problem in detail

I use R/RStudio to draw graphics and maps, and save them as PNG ( with the commands png() and dev.off() ). Along the creation process, I save the same file over and over, to see how it's going, until I get to the final product. Sometimes, the file is fine and I send it to someone or upload somewhere (thus confirming it's ok), just to find out later (minutes, hours, days...) that the file changed somehow. Sometimes they become blank, other times they become one of the previous versions.

However, the modified and the correct file both have exactly the same size, even when the modified is blank. I've posted a more detailed explanation here: https://unix.stackexchange.com/questions/714077/why-are-my-png-files-becoming-corrupt-so-frequently. Although I used the word "corrupt", I can always open the file to see the image, it's just that the image itself changes.

I'm not sure it's a R/RStudio problem, it was suggested in the answers to my question in the link above. I'll pay attention if the changes to the PNG happens while RStudio is still open. Also, it happens so arbitrarily that I cannot use R alone waiting for it to happen — it would be counter productive. I don't think it's a SSD or OS failure, because it happened with two different SSDs and two versions of Debian.

### Describe the behavior you expected

I expected that the PNG files didn't change at all.

- [X] I have read the guide for [submitting good bug reports](https://github.com/rstudio/rstudio/wiki/Writing-Good-Bug-Reports).
- [X] I have installed the latest version of RStudio, and confirmed that the issue still persists.
- [ ] If I am reporting an RStudio crash, I have included a [diagnostics report](https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report).
- [ ] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

**Comment by ronblum** [comment]
@rodrigo1406 Thank you for raising this, and linking to detailed post! Would you be able to provide simple sample code so that I can try to reproduce? Even if it's intermittent, I'd like to try to reproduce it the same way.

**Comment by rodrigo1406** [comment]
I'm sending a folder with the R code (GBIF.R), two shapefiles (BR_UF_2020a and TM_WORLD_BORDERS-0.3), 6 CSVs (one for each plant species), 1 zipped file with an instance of the error (look at Leretia cordata.png and Leretia cordata0.png, both with exactly the same size in bytes) and 2 empty folders, where the new maps will be generated. The exact version of the code was something between maps1 and maps2, but this error happened so many times, with so many different projects, that I don't think it matters. I didn't use the script/files from my linked example, because it uses a very large vegetation raster.

Thanks for your help!

[PNGerror.zip](https://github.com/rstudio/rstudio/files/9453913/PNGerror.zip)

**Comment by rodrigo1406** [comment]
Something new happened today. I made a map (map1.png) and sent it to a friend. Then I made a new version (map2.png), but it wasn't yet ready to be sent. So I planned a new one (map2a.png) to be compared with map2.png. I'm not sure if I really started map2a with png() function, but I know this line was already written, with 'map2a.png' replacing the previous 'map2.png', which in turn replaced the previous 'map1.png' — since I changed the names in the png() command, and since map2.png was already printed as a PNG file, I know I couldn't accidentally overwrite map1.png.

Before printing map2a.png (which I'm not sure if I started with png() function or not), I tried to test some coordinates in the RStudio internal plot area, but I noticed nothing was being printed. So I ran dev.off() line twice: first to close the map2a.png printing (in case it was open) — it said "null device 1" (which I never know if it's an "error" message or not) — then again to be sure nothing else was open — it said "Error in dev.off() : cannot shut down device 1 (the null device)". Since it produced a real error this time, there should be no linking to any of the previous maps, right?

I knew then that I could use the internal plot area to test my coordinates. But then I saw that map1.png received this last "null device 1" plot attempts. How could it be, if I was changing the filenames on the same line? I.e. map2.png, which was the last map, was already complete, as was map1.png before it, so how could RStudio change map1.png? Could it be that my second dev.off() somehow reopened the map1.png connection?

**Comment by rodrigo1406** [comment]
New error. I plotted map3.png and map4.png (again, the line containing 'map3.png' was changed into 'map4.png', so no way to accidentally overwriting map3.png). Both maps were finished. When I closed RStudio, map3.png became identical to map1.png (error described in my previous message), both only a blue square in the top of the image, with the difference that map1.png has only 9.8 kB (error probably produced with the dev.off() above), while map3.png still has its previous 753.3 kB (something went wrong in the very moment when I closed RStudio — I saw the map3.png icon being redrawn in Debian file manager, Nautilus, right after I closed RStudio).

Both files, map1.png and map3.png are attached as a single zip file.

[error maps.zip](https://github.com/rstudio/rstudio/files/9639769/error.maps.zip)

**Comment by rodrigo1406** [comment]
TL;DR

It seems this problem appears when there's an error drawing the image, and/or dev.off() is called an improper number of times (compared to png()). The handle to the file apparently is kept open, leaving to errors when closing the program.

**Comment by kevinushey** [comment]
@rodrigo1406: thank you for the continued follow-up, this is helpful.

Have you had any luck distilling this into a set of reproducible steps? It sounds to me like this can happen if you have multiple graphics devices open at the same time; e.g. something like:

1. Run `png("file1.png")`,
2. Generate a plot,
3. Run `png("file2.png")`,
4. Generate a plot,
5. Close the first graphics device with `dev.off()`,
6. Close the second graphics device with `dev.off()`,

Does this sound accurate? I haven't been able to reproduce following these steps but hopefully we're getting closer.
